x/bridge/keeper/assets_unlocked.go

```diff
@@ -0,0 +1,117 @@
+package keeper
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+
+	"cosmossdk.io/math"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/mezo-org/mezod/x/bridge/types"
+	evmtypes "github.com/mezo-org/mezod/x/evm/types"
+)
+
+// GetAssetsUnlockedSequenceTip returns the current sequence tip for the
+// AssetsUnlocked events. The tip denotes the sequence number of the last event
+// processed by the x/bridge module.
+func (k Keeper) GetAssetsUnlockedSequenceTip(ctx sdk.Context) math.Int {
+	bz := ctx.KVStore(k.storeKey).Get(types.AssetsUnlockedSequenceTipKey)
+
+	var sequenceTip math.Int
+	err := sequenceTip.Unmarshal(bz)
+	if err != nil {
+		panic(err)
+	}
+
+	if sequenceTip.IsNil() {
+		sequenceTip = math.ZeroInt()
+	}
+
+	return sequenceTip
+}
+
+// SetAssetsUnlockedSequenceTip sets the current sequence tip for the AssetsUnlocked
+// events. The tip denotes the sequence number of the last event processed by
+// the x/bridge module.
+func (k Keeper) setAssetsUnlockedSequenceTip(
+	ctx sdk.Context,
+	sequenceTip math.Int,
+) {
+	bz, err := sequenceTip.Marshal()
+	if err != nil {
+		panic(err)
+	}
+
+	ctx.KVStore(k.storeKey).Set(types.AssetsUnlockedSequenceTipKey, bz)
+}
+
+func (k Keeper) saveAssetsUnlocked(
+	ctx sdk.Context,
+	assetsUnlocked *types.AssetsUnlockedEvent,
+) {
+	bz, err := assetsUnlocked.Marshal()
+	if err != nil {
+		panic(err)
+	}
+
+	ctx.KVStore(k.storeKey).Set(types.GetAssetsUnlockedKey(assetsUnlocked.UnlockSequence), bz)
+}
+
+func (k Keeper) GetAssetsUnlocked(
+	ctx sdk.Context,
+	sequence math.Int,
+) *types.AssetsUnlockedEvent {
+	bz := ctx.KVStore(k.storeKey).Get(types.GetAssetsUnlockedKey(sequence))
+
+	var assetsUnlocked types.AssetsUnlockedEvent
+	err := assetsUnlocked.Unmarshal(bz)
+	if err != nil {
+		panic(err)
+	}
+
+	return &assetsUnlocked
+}
+
+func (k Keeper) AssetsUnlocked(
+	ctx sdk.Context,
+	token []byte,
+	amount math.Int,
+	chain uint8,
+	recipient []byte,
+) (*types.AssetsUnlockedEvent, error) {
+	var targetToken common.Address
+	// is it the btc token?
+	btcToken := common.HexToAddress(
+		evmtypes.BTCTokenPrecompileAddress,
+	)
+	if bytes.Equal(btcToken.Bytes(), token) {
+		targetToken = common.BytesToAddress(k.GetSourceBTCToken(ctx))
+	} else {
+		if mapping, ok := k.GetERC20TokenMapping(ctx, token); ok {
+			targetToken = common.BytesToAddress(
+				mapping.SourceTokenBytes(),
+			)
+		}
+	}
+
+	if len(targetToken) == 0 {
+		return nil, fmt.Errorf("unknown token %v", token)
```

Comment:
`token` is `[]bytes`. We should probably encode it to hex before printing:
```
hex.EncodeToString(token)
```
Otherwise, it will be printed as an array of decimal numbers (`[106 20 117 30 ...]`)

Reply:
can do, I avoided to do it for "consistency" as for example with recipient it would not always be valid. Not sure it's a good excuse though.

Reply:
We should probably print all incorrect byte data the same way - either print it in decimal format everywhere or hex-encode it everywhere.
Maybe we could print it everywhere as:
```
return nil, fmt.Errorf("unknown token; hex-encoded value: 0x%x", token)
```
This way we will always know we have incorrect data and we print it in hex-encoded format.
@jeremyletang @lukasz-zimnoch 
WDYT?

Reply:
done here: 6dc7c4ba